File: _125_valid_palindrome/main.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        // THIRD ATTEMPT (with NeetCode video)
        int left = 0;
        int right = s.size() - 1;

        while (left < right) {
            while (left < right && !isalnum(s[left])) {
                left += 1;
            }

            while (right > left && !isalnum(s[right])) {
                right -= 1;
            }

            if (tolower(s[left]) != tolower(s[right])) {
                return false;
            }

            left += 1;
            right -= 1;
        }

        return true;

        // SECOND ATTEMPT: compare with reverse string
        // string str = "";
        // for (char c : s) {
        //     if ((c >= 'a' && c <= 'z') || c >= '0' && c <= '9') {
        //         str += c;
        //     } else if (c >= 'A' && c <= 'Z') {
        //         str += c - 'A' + 'a';
        //     }
        // }

        // return equal(str.rbegin(), str.rend(), str.begin());




        // FIRST ATTEMPT
        // First step: convert to lowercase
        // string str = "";
        // for (char c : s) {
        //     if ((c >= 'a' && c <= 'z') || c >= '0' && c <= '9') {
        //         str += c;
        //     } else if (c >= 'A' && c <= 'Z') {
        //         str += c - 'A' + 'a';
        //     }
        // }

        // // Second step: compare
        // int i = 0;
        // int j = str.size() - 1;

        // while (i < j) {
        //     if (str[i] != str[j]) {
        //         return false;
        //     }
        //     i++;
        //     j--;
        // }

        // return true;
    }
};

File: _125_valid_palindrome/main.py
class Solution:
    def isAlphanumeric(self, c) -> bool:
        if c >= 'a' and c <= 'z':
            return True
        if c >= 'A' and c <= 'Z':
            return True
        if c >= '0' and c <= '9':
            return True

        return False

    def getChar(self, c):
        if c >= 'A' and c <= 'Z':
            return chr(ord(c) - ord('A') + ord('a'))

        return c
    
    def isPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1

        while left < right:
            while left < right and not self.isAlphanumeric(s[left]):
                left += 1
            
            while right > left and not self.isAlphanumeric(s[right]):
                right -= 1

            if self.getChar(s[left]) != self.getChar(s[right]):
                return False

            left += 1
            right -= 1

        return True